docs/transform.ipynb

```diff
@@ -99,15 +137,271 @@
    },
    "outputs": [],
    "source": [
-    "import cirq\n",
-    "c=cirq.Circuit()\n",
-    "c.append(cirq.Moment([]))\n",
-    "c.append(cirq.Moment([cirq.X(cirq.GridQubit(1,1))]))\n",
-    "c.append(cirq.Moment([]))\n",
-    "print(f'Before optimization, Circuit has {len(c)} moments')\n",
+    "@cirq.transformer\n",
+    "def reverse_circuit(circuit: 'cirq.AbstractCircuit', *, context: Optional[cirq.TransformerContext] = None) -> 'cirq.AbstractCircuit':\n",
+    "  \"\"\"Transformer to reverse the input circuit.\"\"\"\n",
+    "  return circuit[::-1]\n",
+    "\n",
+    "\n",
+    "@cirq.transformer\n",
+    "class SubstituteGate:\n",
+    "  \"\"\"Transformer to substitute `source` gates with `target` in the input circuit.\"\"\"\n",
+    "  def __init__(self, source: 'cirq.Gate', target: 'cirq.Gate'):\n",
+    "    self._source = source\n",
+    "    self._target = target\n",
+    "\n",
+    "  def __call__(self, circuit: 'cirq.AbstractCircuit', *, context: Optional[cirq.TransformerContext] = None) -> 'cirq.Circuit':\n",
+    "    batch_replace = []\n",
+    "    for i, op in circuit.findall_operations(lambda op: op.gate == self._source):\n",
+    "      batch_replace.append((i, op, self._target.on(*op.qubits)))\n",
+    "    transformed_circuit = circuit.unfreeze(copy=True)\n",
+    "    transformed_circuit.batch_replace(batch_replace)\n",
+    "    return transformed_circuit\n",
+    "\n",
+    "q = cirq.NamedQubit(\"q\")\n",
+    "circuit = cirq.Circuit(cirq.X(q), cirq.CircuitOperation(cirq.FrozenCircuit(cirq.X(q), cirq.Y(q))), cirq.Z(q))\n",
+    "substitute_gate = SubstituteGate(cirq.X, cirq.S)\n",
+    "print(\"Original Circuit:\", circuit, \"\\n\", sep=\"\\n\")\n",
+    "print(\"Reversed Circuit:\", reverse_circuit(circuit), \"\\n\", sep=\"\\n\")\n",
+    "print(\"Substituted Circuit:\", substitute_gate(circuit), sep=\"\\n\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "OLJrXdyTCGev"
+   },
+   "source": [
+    "## `cirq.TransformerContext` to store common configurable options. \n",
+    "You might be wondering what's the benefit of adding the `@cirq.transformer` decorator on the methods / classes as defined above? Does it add any extra functionality to the decorated objects at all? \n",
+    "\n",
+    "Yes! The transformer decorator inspects the `cirq.TransformerContext` argument and automatically appends useful functionality like support for automated logging and recursively running the transformer on nested sub-circutis. \n"
```

Comment:
Nit: generally philosphy is to instruct the reader and not sell/editorialize where possible.

Reply:
reworded.